src/lib/storage/lz4_segment/lz4_segment_iterable.hpp

```diff
@@ -19,18 +19,15 @@ class LZ4SegmentIterable : public PointAccessibleSegmentIterable<LZ4SegmentItera
   template <typename Functor>
   void _on_with_iterators(const Functor& functor) const {
     using ValueIterator = typename std::vector<T>::const_iterator;
+    using NullValueIterator = typename pmr_vector<bool>::const_iterator;
 
     auto decompressed_segment = _segment.decompress();
 
-    /**
-     * If the null value vector doesn't exist, then the segment does not have any row value that is null. In that case,
-     * we can just use a default initialized boolean vector.
-     */
-    const auto null_values = _segment.null_values() ? *_segment.null_values() : pmr_vector<bool>(_segment.size());
-
-    auto begin = Iterator<ValueIterator>{decompressed_segment.cbegin(), null_values.cbegin()};
-    auto end = Iterator<ValueIterator>{decompressed_segment.cend(), null_values.cend()};
-
+    std::optional<NullValueIterator> null_opt_value_iter;
+    auto begin = Iterator<ValueIterator>{
+        decompressed_segment.cbegin(), _segment.null_values() ? _segment.null_values()->cbegin() : null_opt_value_iter};
```

Comment:
Will the `null_values` iterator be copied here in case it exists?

Reply:
You create it here as part of the method call, so it's a prvalue. Currently, you copy it from the `const&` to the local member as part of your iterator constructor. You could avoid that by accepting it by value and moving it into the member.

Reply:
Is there a way around that error?
```../src/lib/storage/lz4_segment/lz4_segment_iterable.hpp:26:63: error: incompatible operand types ('std::__1::vector<bool, boost::container::pmr::polymorphic_allocator<bool> >::const_iterator' (aka '__bit_iterator<std::__1::vector<bool, boost::container::pmr::polymorphic_allocator<bool> >, true>') and 'const std::__1::nullopt_t')
        decompressed_segment.cbegin(), _segment.null_values() ? _segment.null_values()->cbegin() : std::nullopt};```

that is not an additional wrap like this `std::optional<NullValueIterator>{_segment.null_values()}`????

And is the move really necessary (adding any advantage) here?

Reply:
Instead of `std::nullopt`, try to use `std::optional<NullValueIterator>{}`. Sorry.

> And is the move really necessary (adding any advantage) here?

No idea, but when a constructor takes a complex object by value, moving it is more or less the standard.

Reply:
No worries. As often as I am positively surprised what compilers are capable of, as often I am totally confused why simple things cannot be resolved. Will add the move.

Reply:
Changed. Declaring the type with the nullopt was shorter than what I had before.